### PR TITLE
Experiment with separating concerns in some simple States.

### PR DESCRIPTION
**Pull request by dabreegster**
This is closer to a callback style, which might be more familiar to
people coming from other frameworks.

It's a pretty mechanical refactor now, but the hope is that looking at a
bunch of on_mouseover() functions will reveal further refactors more
easily than a bunch of monolithic event() functions.

I don't think every State fits this pattern; things like SandboxMode
need to carefully orchestrate the order of things. Make simple things
easy and hard things possible.

@michaelkirk, @gedkott : Any opinion on this? There are many more places it could be applied, but I don't want to spend the time before being more convinced this is helpful.

### CODE REVIEW

File: game/src/common/mod.rs [hunk 1]
```diff
@@ -410,3 +411,60 @@ pub fn checkbox_per_mode(
     }
     Widget::custom_row(filters)
 }
+
+/// Many game states fit a pattern of managing a single panel, handling mouseover events, and other
+/// interactions on the map. Implementing this instead of `State` reduces some boilerplate.
+pub trait SimpleState {
+    /// Called when something on the panel has been clicked.
+    fn on_click(&mut self, ctx: &mut EventCtx, app: &mut App, x: &str, panel: &Panel)
```
Comment (michaelkirk) [inline game/src/common/mod.rs:419]: Do we intentionally not have a default impl for `on_click`?

File: game/src/common/mod.rs [hunk 2]
```diff
@@ -410,3 +411,60 @@ pub fn checkbox_per_mode(
...
+        -> Transition;
+    /// Called when something on the panel has changed.
+    fn panel_changed(&mut self, _: &mut EventCtx, _: &mut App, _: &Panel) -> Option<Transition> {
+        None
+    }
+    /// Called when the mouse has moved.
+    fn on_mouseover(&mut self, _: &mut EventCtx, _: &mut App) {}
+    /// Always called if the previous methods have not terminated.
```
Comment (michaelkirk) [inline game/src/common/mod.rs:427]: What does "terminated" mean in this context? 

It looks to me like we'll call `other_event` regardless of what happens in `on_mouseover`, so maybe referring to "the previous methods" should be "if `on_click or `panel_changed`"

**Review by michaelkirk — APPROVED** [review]
Nice work - I think this more opinionated API will be helpful for implementors. 👍

**Comment by michaelkirk** [comment]
> There are many more places it could be applied, but I don't want to spend the time before being more convinced this is helpful.

As a side note, if SimpleState becomes more popular than State we should consider changing the names  - maybe SimpleState becomes State and what *was* State becomes CustomState or something.

File: game/src/common/mod.rs (hunk 1 above)
Comment (dabreegster) [reply game/src/common/mod.rs:419]: There's no reasonable default (except maybe treating "close" as `Transition::Pop`, but that's a bit too much implicit convention). At minimum, a `SimpleState` only has a panel with different actions, and no real map-space handling. I updated the comments a bit to make the usage more clear.

File: game/src/common/mod.rs (hunk 2 above)
Comment (dabreegster) [reply game/src/common/mod.rs:427]: Good point! Updated the comment

**Comment by dabreegster** [comment]
> As a side note, if SimpleState becomes more popular than State we should consider changing the names - maybe SimpleState becomes State and what was State becomes CustomState or something.

Quite reasonable! I'm not sure how many places should be changed over; I don't think I'll make a focused effort to migrate, but just do it as I happen to touch different parts of the UI.

**Pull request merged**